runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,212 @@
+      range.s[1] + 1,
+      range.s[2] + 1,
+      range.e[1] + 1,
+      range.e[2]
+    ), 'in', true)
+  end
+end
+
+---Adjust snippet indent
+---@param snippet_text string
+---@return string
+local function adjust_indent(snippet_text)
+  local one_indent = get_one_indent()
+  local base_indent = get_base_indent()
+  local adjusted_snippet_text = {}
+  local texts = vim.split(snippet_text, '\n', true)
+  for i, text in ipairs(texts) do
+    if i ~= #texts and string.match(text, '^%s*$') then
+      table.insert(adjusted_snippet_text, '')
+    else
+      if i ~= 1 then
+        text = string.gsub(text, '^(\t+)', function(indent)
+          return string.gsub(indent, '\t', one_indent)
+        end)
+        table.insert(adjusted_snippet_text, base_indent .. text)
+      else
+        table.insert(adjusted_snippet_text, text)
+      end
+    end
+  end
+  return table.concat(adjusted_snippet_text, '\n')
+end
+
+---@alias nvim.SnippetJumpDirection "1" | "2"
+local JumpDirection = {}
+JumpDirection.Next = 1
+JumpDirection.Prev = 2
+
+
+---@class nvim.SnippetSession
+---@field ns string
+---@field public snippet_text string
+local SnippetSession = {}
+
+---Create SnippetSession instance.
+---@param snippet_text string
+function SnippetSession.new(snippet_text)
+  return setmetatable({
+    ns = vim.api.nvim_create_namespace('nvim.snippet'),
+    snippet_text = adjust_indent(snippet_text),
+  }, { __index = SnippetSession })
+end
+
+---Insert snippet to the current buffer.
+function SnippetSession:expand()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  cursor[1] = cursor[1] - 1
+
+  local snippet_node = snippet.parse(self.snippet_text)
+  vim.api.nvim_buf_set_text(0, cursor[1], cursor[2], cursor[1], cursor[2], vim.split(tostring(snippet_node), '\n'))
+
+  local function traverse(node, start_pos)
+    local end_pos = { start_pos[1], start_pos[2] }
+    local marks = {}
+    if node.children then
+      for _, child in ipairs(node.children) do
+        for _, mark in ipairs(traverse(child, end_pos)) do
+          table.insert(marks, mark)
+        end
+      end
+    else
+      local texts = vim.split(tostring(node), '\n')
+      end_pos[1] = start_pos[1] + #texts - 1
+      end_pos[2] = #texts == 1 and (start_pos[2] + #texts[1]) or #texts[#texts]
+    end
+
+    if vim.tbl_contains({
+      snippet.NodeType.TABSTOP,
+      snippet.NodeType.PLACEHOLDER,
+      snippet.NodeType.CHOICE,
+    }, node.type) then
+      table.insert(marks, {
+        mark_id = vim.api.nvim_buf_set_extmark(0, self.ns, start_pos[1], start_pos[2], {
+          end_line = end_pos[1],
+          end_col = end_pos[2],
+          hl_group = 'Visual',
+        }),
+        node = node,
+      })
+    end
+
+    start_pos[1] = end_pos[1]
+    start_pos[2] = end_pos[2]
+
+    return marks
+  end
+  self.marks = traverse(snippet_node, cursor)
+end
+
+---Jump to the suitable placeholder.
+---@param direction nvim.SnippetJumpDirection
+function SnippetSession:jump(direction)
```

Comment:
This implementation does not care about the `current tabstop index`. Instead of that, this determines the suitable next tabstop via `cursor position`.

This is not general implementation. I think we should implement correct version if we ship this PR.